### PR TITLE
Add the support of the cross-compiling for iOS devices

### PR DESCRIPTION
**Pull request by Xreki**
Xcode is used as the toolchain for cross-compiling for iOS devices. Make sure Xcode is installed on your Mac.

Use the following `CMake` command
```
cmake -DCMAKE_INSTALL_PREFIX=/home/liuyiqun/paddle \
      -DCMAKE_SYSTEM_NAME=iOS \
      -DIOS_PLATFORM=OS \
      -DCMAKE_OSX_ARCHITECTURES="arm64" \
      -DWITH_C_API=ON \
      -DUSE_EIGEN_FOR_BLAS=ON \
      -DWITH_TESTING=OFF \
      -DWITH_SWIG_PY=OFF \
      -DWITH_STYLE_CHECK=OFF \
      -DCMAKE_BUILD_TYPE=Release \
      ..
make
make install
```

Set `-DIOS_USE_VECLIB_FOR_BLAS=ON` to use `veclib`, such as
```
cmake -DCMAKE_INSTALL_PREFIX=/home/liuyiqun/paddle \
      -DCMAKE_SYSTEM_NAME=iOS \
      -DIOS_PLATFORM=OS \
      -DCMAKE_OSX_ARCHITECTURES="arm64" \
      -DIOS_USE_VECLIB_FOR_BLAS=ON \
      -DWITH_C_API=ON \
      -DWITH_TESTING=OFF \
      -DWITH_SWIG_PY=OFF \
      -DWITH_STYLE_CHECK=OFF \
      -DCMAKE_BUILD_TYPE=Release \
      ..
```

When neither `IOS_USE_VECLIB_FOR_BLAS` nor `USE_EIGEN_FOR_BLAS` is set , `OpenBLAS` will be used. You can change `CMAKE_OSX_ARCHITECTURES` to `"armv7"`, but note `OpenBLAS` can not be built for `armv7`.

TODO:
- Support multiple architectures through a single `make` process
- Fix the linking problem

### CODE REVIEW

**Comment by NHZlX** [comment]
look forward to the support for armv7 and simulator

**Review by hedaoyuan — DISMISSED** [review]
赞！

File: CMakeLists.txt [hunk 1]
```diff
@@ -64,27 +64,27 @@ if(NOT CMAKE_BUILD_TYPE)
       FORCE)
 endif()
 
-if(ANDROID)
-    if(${CMAKE_SYSTEM_VERSION} VERSION_LESS "16")
+if(ANDROID OR IOS)
+    if(ANDROID AND ${CMAKE_SYSTEM_VERSION} VERSION_LESS "16")
```
[the remark this answers is not in the document]
Comment (Xreki) [reply CMakeLists.txt]: 对的，merge的时候没注意，我下一个pr改过来吧。

File: cmake/external/openblas.cmake [hunk 2]
```diff
@@ -29,37 +29,50 @@ IF(NOT ${CBLAS_FOUND})
...
-    ENDIF()
+    SET(OPENBLAS_CC "${CMAKE_C_COMPILER}")
 
     IF(CMAKE_CROSSCOMPILING)
+        SET(OPTIONAL_ARGS HOSTCC=${HOST_C_COMPILER})
+        GET_FILENAME_COMPONENT(CROSS_SUFFIX ${CMAKE_C_COMPILER} DIRECTORY)
+        SET(CROSS_SUFFIX ${CROSS_SUFFIX}/)
         IF(ANDROID)
             # arm_soft_fp_abi branch of OpenBLAS to support softfp
             #   https://github.com/xianyi/OpenBLAS/tree/arm_soft_fp_abi
```
[the remark this answers is not in the document]
Comment (Xreki) [reply cmake/external/openblas.cmake]: 这个分支的修改内容似乎并没有合入`master`或者`develop`分支，从两个分支的`Makefile.arm`看的出来：
- [Makefile.arm on arm_soft_fp_abi](https://github.com/xianyi/OpenBLAS/blob/arm_soft_fp_abi/Makefile.arm#L14)
```
ifeq ($(CORE), ARMV7)
ifeq ($(OSNAME), Android)
ifeq ($(ARM_SOFTFP_ABI), 1)
CCOMMON_OPT += -mfpu=neon -march=armv7-a
FCOMMON_OPT += -mfpu=neon -march=armv7-a
else
CCOMMON_OPT += -mfpu=neon -march=armv7-a -Wl,--no-warn-mismatch
FCOMMON_OPT += -mfpu=neon -march=armv7-a -Wl,--no-warn-mismatch
endif
else
CCOMMON_OPT += -mfpu=vfpv3 -march=armv7-a
FCOMMON_OPT += -mfpu=vfpv3 -march=armv7-a
endif
endif
```
- [Makefile.arm on develop](https://github.com/xianyi/OpenBLAS/blob/develop/Makefile.arm)

[the remark this answers is not in the document]
Comment (hedaoyuan) [reply cmake/external/openblas.cmake]: `arm_soft_fp_abi`分支只显示`This branch is 242 commits behind develop.`说明它已经merge到develop了，否则应该会显示commits ahead of develop这类的。至于，develop的Makefile.arm与arm_soft_fp_abi的不一样可能是develop后来又修改了。
ps，我用develop的编译跑了一下，好像没发现比arm_soft_fp_abi慢。

[the remark this answers is not in the document]
Comment (Xreki) [reply cmake/external/openblas.cmake]: 我验证一下，下一个PR修复吧。

**Review by hedaoyuan — APPROVED** [review]

**Pull request merged**